From: Ivan Vecera <ivecera@redhat.com>
Date: Thu, 22 Jul 2010 14:10:42 -0400
Subject: [net] r8169: fix for broken register writes
Message-id: <1279807842-10377-1-git-send-email-ivecera@redhat.com>
Patchwork-id: 27035
O-Subject: [RHEL5.6 PATCH] r8169: fix for broken register writes
Bugzilla: 581654
RH-Acked-by: Jarod Wilson <jarod@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>

BZs:
#581654 - RTL-8169 Gigatit Ethernet network devices mac address changes after soft reboot

Description:
This patch fixes a problem with register writes (MAC address and
multicast). These registers needs to be written in reverse order
(1st high 32 bits and then low 32 bits).

Test status:
Tested by myself and by bug reporter

Upstream commits:
78f1cd0 - r8169: fix broken register writes
908ba2b - r8169: more broken register writes workaround

Signed-off-by: Ivan Vecera <ivecera@redhat.com>

diff --git a/drivers/net/r8169.c b/drivers/net/r8169.c
index df7f5b4..1a16d23 100644
--- a/drivers/net/r8169.c
+++ b/drivers/net/r8169.c
@@ -2832,8 +2832,13 @@ static void rtl_rar_set(struct rtl8169_private *tp, u8 *addr)
 	spin_lock_irq(&tp->lock);
 
 	RTL_W8(Cfg9346, Cfg9346_Unlock);
-	RTL_W32(MAC0, low);
+
 	RTL_W32(MAC4, high);
+	RTL_R32(MAC4);
+
+	RTL_W32(MAC0, low);
+	RTL_R32(MAC0);
+
 	RTL_W8(Cfg9346, Cfg9346_Lock);
 
 	spin_unlock_irq(&tp->lock);
@@ -4830,8 +4835,8 @@ static void rtl_set_rx_mode(struct net_device *dev)
 		mc_filter[1] = swab32(data);
 	}
 
-	RTL_W32(MAR0 + 0, mc_filter[0]);
 	RTL_W32(MAR0 + 4, mc_filter[1]);
+	RTL_W32(MAR0 + 0, mc_filter[0]);
 
 	RTL_W32(RxConfig, tmp);
 
